### PR TITLE
Support for the Youtube syntax

### PR DESCRIPTION
**Pull request by takker99**
## Proposed Changes

Support for the Youtube syntax

### CODE REVIEW

**Comment by takker99** [comment]
Codeclimate still warn an issue about a cognitive complexity.
However I'm not sure how to refactor the code, so I ignore it, sorry.

**Comment by takker99** [comment]
Besides a codeclimate issue, there is a unit test error I can't understand.

```sh
$ npx jest 
 PASS  tests/utils/parseParams.test.ts (6.251 s)
 PASS  tests/line/link.test.ts (6.302 s)
 PASS  tests/line/image.test.ts (6.254 s)
 PASS  tests/line/youtube.test.ts (6.244 s)
 PASS  tests/codeBlock/index.test.ts (6.272 s)
 PASS  tests/page/index.test.ts (6.321 s)
 PASS  tests/line/strongImage.test.ts
 PASS  tests/line/helpfeel.test.ts
 PASS  tests/line/bullet.test.ts
 PASS  tests/table/index.test.ts
 PASS  tests/line/blank.test.ts
 PASS  tests/line/commandLine.test.ts
 PASS  tests/line/icon.test.ts
 PASS  tests/line/formula.test.ts
 PASS  tests/line/quote.test.ts
 PASS  tests/line/strongIcon.test.ts
 PASS  tests/line/code.test.ts
 PASS  tests/line/strong.test.ts
 PASS  tests/line/googleMap.test.ts
 PASS  tests/line/hashTag.test.ts     
 PASS  tests/line/index.test.ts
 PASS  tests/line/plain.test.ts
 FAIL  tests/line/decoration.test.ts
  ● Test suite failed to run

    tests/line/decoration.test.ts:34:12 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    34 [~ deco]`).toMatchSnapshotWhenParsing({ hasTitle: false });
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:68:33 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    68     expect("[*********** 11*]").toMatchSnapshotWhenParsing({ hasTitle: false });
                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:72:44 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    72     expect("[* hoge https://example.com]").toMatchSnapshotWhenParsing({
                                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:78:27 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    78     expect("[! [[[[[[a]").toMatchSnapshotWhenParsing({ hasTitle: false });
                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:82:28 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    82     expect("[! [[[[[[a]]").toMatchSnapshotWhenParsing({ hasTitle: false });
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:86:28 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    86     expect("[* [[link]]]").toMatchSnapshotWhenParsing({ hasTitle: false });
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:90:36 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    90     expect("[* [[progfay.icon]]]").toMatchSnapshotWhenParsing({
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:96:53 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    96     expect("[* [[https://example.com/image.png]]]").toMatchSnapshotWhenParsing({
                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~

 PASS  tests/title/index.test.ts (7.416 s)

Summary of all failing tests
 FAIL  tests/line/decoration.test.ts
  ● Test suite failed to run

    tests/line/decoration.test.ts:34:12 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    34 [~ deco]`).toMatchSnapshotWhenParsing({ hasTitle: false });
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:68:33 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    68     expect("[*********** 11*]").toMatchSnapshotWhenParsing({ hasTitle: false });
                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:72:44 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    72     expect("[* hoge https://example.com]").toMatchSnapshotWhenParsing({
                                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:78:27 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    78     expect("[! [[[[[[a]").toMatchSnapshotWhenParsing({ hasTitle: false });
                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:82:28 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    82     expect("[! [[[[[[a]]").toMatchSnapshotWhenParsing({ hasTitle: false });
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:86:28 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    86     expect("[* [[link]]]").toMatchSnapshotWhenParsing({ hasTitle: false });
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:90:36 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    90     expect("[* [[progfay.icon]]]").toMatchSnapshotWhenParsing({
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~
    tests/line/decoration.test.ts:96:53 - error TS2339: Property 'toMatchSnapshotWhenParsing' does not exist on type 'JestMatchers<string>'.

    96     expect("[* [[https://example.com/image.png]]]").toMatchSnapshotWhenParsing({
                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~


Test Suites: 1 failed, 23 passed, 24 total
Tests:       105 passed, 105 total
Snapshots:   92 passed, 92 total
Time:        8.025 s, estimated 9 s
Ran all test suites.
```

The type of `toMatchSnapshotWhenParsing` is already defined at `tests/jest-setup.ts`. Nevertheless Jest complain such a type is not defined anywhere. The error confuses me. @progfay , what do you think a cause of the error?

File: src/block/node/index.ts [hunk 1]
```diff
@@ -59,6 +60,7 @@ export const convertToNodes = combineNodeParsers(
   StrongImageNodeParser,
   StrongIconNodeParser,
   StrongNodeParser,
+  YoutubeNodeParser,
```
Comment (takker99) [inline src/block/node/index.ts:63]: When we remove the line, the jest error is suppressed.

**Comment by progfay** [comment]
I seem that this jest error was caused by cache.

Please try:
```sh
npm run test -- --clearCache
npm test
```

**Comment by progfay** [comment]
scrapbox.io supports YouTube and parser of scrapbox.io create YouTube node.

However, this project does not plan to support each media Node. (such as YouTube, Twitter, audio)
You can do this by checking `LinkNode.href` if you want.

**Comment by takker99** [comment]
I'm sorry having written the reply and close the PR by myself.
I understand your plan. I'm going to parse such Nodes by checking `LinkNode.href`.

> I think that this jest error is caused by cache.
> 
> Please try:
> 
> ```shell
> npm run test -- --clearCache
> npm test
> ```
Thanks a lot! I resolved the error by running the command!


By the way, do you mind if I asked you why you wouldn't implemented each media Node?
Even if it makes sense not to implemented media Nodes like Youtube and Vimeo, I wonder why scrapbox-parser don't plan to support audio Node (and video Node), though this already implemented image Node.

Of course image Node is more frequently used than audio and video Nodes, so it's natural to parse image Node independently. I'm merely curious about the question. You don't need to be serious.

**Comment by progfay** [comment]
Image node is base feature of scrapbox.io, this feature will never be removed.

But, media type node feature may be added/removed in the future.
It is expensive to implement with each time, so not supported.

Another reason is that we believe that the fewer types of Nodes, the easier it is for users to handle.

**Pull request closed without merging**